File: golang-service/internal/metrics/metrics.go
package metrics

import (
	"fmt"
	"net/http"
	"runtime"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	httpRequestsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "http_requests_total",
			Help: "Total number of HTTP requests",
		},
		[]string{"method", "endpoint", "status"},
	)

	httpRequestDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "http_request_duration_seconds",
			Help:    "HTTP request duration in seconds",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"endpoint"},
	)

	appMemoryBytes = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "app_memory_bytes",
			Help: "Application memory usage in bytes",
		},
	)

	appCPUSecondsTotal = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "app_cpu_seconds_total",
			Help: "Total CPU time used by the application (placeholder)",
		},
	)

	dbLockWaitSecondsTotal = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "db_lock_wait_seconds_total",
			Help: "Total time spent waiting for database locks",
		},
	)
)

var registry *prometheus.Registry

// Инициализация системы метрик Prometheus
func Init() {
	registry = prometheus.NewRegistry()
	registry.MustRegister(httpRequestsTotal)
	registry.MustRegister(httpRequestDuration)
	registry.MustRegister(appMemoryBytes)
	registry.MustRegister(appCPUSecondsTotal)
	registry.MustRegister(dbLockWaitSecondsTotal)

	// Запуск фоновой горутины профилирования потребления памяти
	go updateMemoryMetrics()
}

// Забор статистики потребления оперативной памяти
func updateMemoryMetrics() {
	ticker := time.NewTicker(5 * time.Second)
	// Отложенная остановка таймера
	defer ticker.Stop()

	for range ticker.C {
		var m runtime.MemStats
		// Чтение состояния распределения памяти
		runtime.ReadMemStats(&m)
		appMemoryBytes.Set(float64(m.Alloc))
	}
}

// Интеграция промежуточного слоя сбора HTTP-статистики
func MetricsMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		rw := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}

		next.ServeHTTP(rw, r)

		duration := time.Since(start).Seconds()
		endpoint := normalizePath(r.URL.Path)

		statusCode := fmt.Sprintf("%d", rw.statusCode)
		// Увеличение счетчика входящих запросов
		httpRequestsTotal.WithLabelValues(r.Method, endpoint, statusCode).Inc()
		// Наблюдение за задержками времени ответа
		httpRequestDuration.WithLabelValues(endpoint).Observe(duration)
	})
}

// Структура перехватчика HTTP ответов для захвата статус-кода
type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

// Захват записи заголовка ответа
func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

// Возврат HTTP обработчика модуля Prometheus
func Handler() http.Handler {
	return promhttp.HandlerFor(registry, promhttp.HandlerOpts{})
}

// Регистрация времени ожидания блокировок СУБД
func RecordDBLockWait(duration time.Duration) {
	dbLockWaitSecondsTotal.Add(duration.Seconds())
}

File: python-service/internal/metrics/metrics.py
import os
import psutil
import asyncio
import time
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response
from prometheus_client import Counter, Histogram, Gauge, CollectorRegistry

# Нативная реестровая запись для Prometheus
registry = CollectorRegistry()

# Объявление метрик HTTP запросов
http_requests_total = Counter(
    "http_requests_total",
    "Total number of HTTP requests",
    ["method", "endpoint", "status"],
    registry=registry
)

http_request_duration = Histogram(
    "http_request_duration_seconds",
    "HTTP request duration in seconds",
    ["endpoint"],
    registry=registry
)

# Объявление метрик аппаратных ресурсов
app_memory_bytes = Gauge(
    "app_memory_bytes",
    "Application memory usage in bytes",
    registry=registry
)

app_cpu_seconds_total = Counter(
    "app_cpu_seconds_total",
    "Total CPU time used by the application (placeholder)",
    registry=registry
)

# Метрики ожидания транзакций
db_lock_wait_seconds_total = Counter(
    "db_lock_wait_seconds_total",
    "Total time spent waiting for database locks",
    registry=registry
)

def record_db_lock_wait(duration_seconds: float) -> None:
    """Регистрация времени ожидания блокировок СУБД."""
    db_lock_wait_seconds_total.inc(duration_seconds)

async def update_memory_metrics() -> None:
    """Забор статистики потребления оперативной памяти."""
    process = psutil.Process(os.getpid())
    while True:
        try:
            # Чтение состояния распределения памяти
            app_memory_bytes.set(process.memory_info().rss)
        except Exception:
            pass
        await asyncio.sleep(5)

class MetricsMiddleware(BaseHTTPMiddleware):
    """Интеграция промежуточного слоя сбора HTTP-статистики."""
    
    async def dispatch(self, request: Request, call_next) -> Response:
        start_time = time.time()
        endpoint = request.url.path

        try:
            # Делегирование запроса слоям бизнес-логики
            response = await call_next(request)
            status_code = response.status_code
        except Exception as e:
            status_code = 500
            raise e
        finally:
            duration = time.time() - start_time
            
            # Увеличение счетчика входящих запросов
            http_requests_total.labels(
                method=request.method, 
                endpoint=endpoint, 
                status=str(status_code)
            ).inc()
            
            # Наблюдение за задержками времени ответа
            http_request_duration.labels(endpoint=endpoint).observe(duration)
            
        return response
